libminifi/include/core/Repository.h

```diff
@@ -218,56 +195,77 @@ class Repository : public virtual core::SerializableComponent, public core::Trac
     return Put(key, buffer, bufferSize);
   }
 
-  uint64_t incrementSize(const char* /*fpath*/, const struct stat *sb, int /*typeflag*/) {
-    return (repo_size_ += sb->st_size);
-  }
-
   virtual void loadComponent(const std::shared_ptr<core::ContentRepository>& /*content_repo*/) {
   }
 
-  virtual uint64_t getRepoSize();
+  virtual uint64_t getRepoSize() const {
+    return repo_size_;
+  }
 
   std::string getDirectory() const {
     return directory_;
   }
 
-  // Prevent default copy constructor and assignment operation
-  // Only support pass by reference or pointer
-  Repository(const Repository &parent) = delete;
-  Repository &operator=(const Repository &parent) = delete;
+  // Starts repository monitor thread
+  void start() {
+    if (running_.exchange(true)) {
+      return;
+    }
+    if (purge_period_ <= std::chrono::milliseconds(0)) {
+      return;
+    }
+    getThread() = std::thread(&Repository::run, this);
+
+    logger_->log_debug("%s Repository monitor thread start", name_);
+  }
+
+  // Stops repository monitor thread
+  void stop() {
+    if (!running_.exchange(false)) {
+      return;
+    }
+    if (getThread().joinable()) {
+      getThread().join();
+    }
+    logger_->log_debug("%s Repository monitor thread stop", name_);
+  }
```

Comment:
It's possible to start again on another thread after line 224 has finished, and we're waiting for 227 or 228. One should not be able to start the repository until stopping has finished. Same with start, you shouldn't be able to stop until starting has finished.
I suggest using a state enum with 4 values: starting, running, stopping, stopped.
Only transitions to the right and stopped -> starting should be possible.

Reply:
Done.